Upgrade aws-sdk to version 3

agrubi opened the pull request:
Upgrade aws-sdk to version 3
This is a breaking change, it requires clients to configure dynogels with aws-sdk v3.

- set service property of documentClient to `DynamoDb` instance. That's what the code expects.
- Change Set handling to just use native Set instead of some custom implementation. This is now possible, because aws-sdk 3 has support for sets. 
- Replace buffer generation with an aws-sdk v3 version of it.
- Adjust tests that expected error.code to expect error.name instead (breaking change in aws-sdk 3)
- Adjust tests that expected Projection to always expect NonKeyAttributes value (change in aws-sdk 3)
- Remove AWS config from examples and readme

Made sure all unit tests pass

![persistence](https://media.giphy.com/media/Rio1R7xCj3N1IaocRy/giphy.gif)

agrubi committed:
Upgrade aws-sdk to version 3

- set service property of documentClient to `DynamoDb` instance. That's what the code expects.
- Change Set handling to just use native Set instead of some custom implementation. This is now possible, because aws-sdk 3 has support for sets. It's a breaking change, I think. Clients should migrate their models to use native sets (just search for `stringSet()`, 'numberSet()', 'binarySet()' and `dynamoType` in your code base to see whether you're using sets)
- Replace buffer generation with an aws-sdk v3 version of it.
- Adjust tests that expected error.code to expect error.name instead (breaking change in aws-sdk 3)
- Adjust tests that expected Projection to always expect NonKeyAttributes value (change in aws-sdk 3)
- Remove AWS config from examples and readme

Made sure all unit tests pass

agrubi committed:
Optionally accept documentClient in dynamoDriver()

This allows passing in a customized documentClient (e.g. `DynamoDBDocument.from(exports.dynamodb, { marshallOptions: { removeUndefinedValues: true } });`)

agrubi committed:
Change aws-sdk dependency to be more permissive

agrubi committed:
Preserve previous behaviour for Sets and Binary

Sets used to be deserialized into an array, keep it that way
Binary fields used to be deserialized into Binary object, keep it that way (also for binary arrays and sets)
Add integration test for binary fields

ThomasMorrison commented on package.json:
```diff
@@ -27,12 +27,14 @@
   },
   "license": "MIT",
   "dependencies": {
+    "@aws-sdk/client-dynamodb": "^3.0.0",
+    "@aws-sdk/lib-dynamodb": "^3.0.0",
+    "@aws-sdk/util-buffer-from": "^3.0.0",
```
Should these remain peer dependencies?

My main concern is API upgrading one of these, and then passing in some object/client which has new behaviour to a dynogels constructor.

agrubi replied:
Changed to peer dependency. This allows us to fail in cases where the library user has non-matching aws version (e.g. version 4 or 2).

agrubi committed:
Change aws to be a peer dependency

This allows us to fail in cases where the library user has non-matching aws version (e.g. version 4 or 2).

ThomasMorrison reviewed (approved):
:cake:
